dcpy/extract/ingest.py

```diff
@@ -1,30 +1,87 @@
 
+from dcpy.utils.logging import logger
+from dcpy.connectors.edm import recipes, publishing
+from . import TMP_DIR, PARQUET_PATH, metadata, utils
 
-# doesn't strictly need to be recipe config, but at least config-like object
-def archive_raw_dataset(import_config: config.Import):
-    """Given dataset and potentially a semantic version, ingest it from its source and archive it in its
-    raw format, with metadata, in edm-recipes/raw_data"""
-    ## logic to grab dataset based on ImportDefinition, dump locally
 
-    ## logic to "compute" recipes.Config object, similar to Config.compute in library
-    ## should be limited to metadata gathering - version, etc
-    ## leaving here as opposed to helper function because not sure if it will only depend on import definition
-    ##   or read any data from downloaded dataset
+def _download_file_web(url: str, file_name: str) -> Path:
+    """Simple wrapper to download a file using requests.get.
+    Returns filepath."""
+    TMP_DIR.mkdir(exist_ok=True)
+    file = TMP_DIR / file_name
+    logger.info(f"downloading {url} to {file}")
+    response = requests.get(url)
+    response.raise_for_status()
+    with open(file, "wb") as f:
+        f.write(response.content)
+    return file
 
-    ## logic to push local dataset and "computed" config to raw archive in s3
 
-    ## return Config, from which s3path should be computed
+def download_file_from_source(template: metadata.Template, version: str) -> Path:
+    """From parsed config template and version, download raw data from source
+    and return path of saved local file."""
+    match template.source:
+        ## Non reqeust-based methods
+        case recipes.ExtractConfig.Source.LocalFile() as local_file:
+            return local_file.path
+        case recipes.ExtractConfig.Source.EdmPublishingGisDataset() as dataset:
+            return publishing.download_gis_dataset(dataset.name, version, TMP_DIR)
+        case recipes.ExtractConfig.Source.Script() as s:
+            module = importlib.import_module(
+                f"{s.script_module or 'dcpy.extract.ingest_scripts'}.{s.script_name or template.name}"
+            )
+            logger.info(f"Running custom ingestion script {template.name}.py")
+            return module.runner()
 
-    raise NotImplemented
+        ## request-based methods
+        case recipes.ExtractConfig.Source.FileDownload() as file_download:
+            return _download_file_web(
+                file_download.url, os.path.basename(urlparse(file_download.url).path)
+            )
+        case recipes.ExtractConfig.Source.Api() as api:
+            return _download_file_web(api.endpoint, f"{template.name}.{api.format}")
+        case recipes.ExtractConfig.Source.Socrata() as socrata:
+            return _download_file_web(
+                utils.Socrata.get_url(socrata), f"{template.name}.{socrata.extension}"
+            )
+
+
+def archive_raw_dataset(dataset: str, version: str | None):
```

Comment:
This needs a better name, it shouldn't have the same name as the recipes function which does the actual archiving. Something that better represents the scope of this - get data from source, determine metadata, archive, and return config/metadata. Please help

Reply:
copilot recs for naming: 
* `get_data_archive_and_return_config`
* `retrieve_raw_data_and_get_metadata`